**<h1> Accelerating Insights: A Data-Driven Look at Vehicle Fuel Efficiency and Design </h1>**
#### Done By: Lamel Kekana

<div align="center" style=" font-size: 30%; text-align: center; margin: 0 auto">
<img src="https://cdn.thezebra.com/zfront/media/production/images/8xzyovqA.width-800.jpg" width="800" height="400" />
</div>

---

<h3 id="cont">Table of Contents</h1>

<a href="#BC"> Background Context</a>

<a href="#one">1. Importing Packages</a>

<a href="#two">2. Data Collection and Description</a>

<a href="#three">3. Data Loading and Inspection </a>

<a href="#four">4. Data Cleaning and Filtering</a>

<a href="#five">5. Exploratory Data Analysis (EDA)</a>

<a href="#six">6. Modeling </a>

<a href="#seven">7. Evaluation and Validation</a>

<a href="#eight">8. Final Model</a>

<a href="#nine">9. Conclusion and Future Work</a>

<a href="#ten">10. References</a>

---

## **Background Context**
<a id="BC"></a>
<a href=#cont>Back to Table of Contents</a>


#### __*Introduction*__


This dataset provides a comprehensive view of vehicle specifications and performance metrics, offering valuable insights into fuel efficiency, engine characteristics, and design trends across various car models and manufacturers. It includes information about fuel consumption in city, highway, and combined driving conditions, alongside key attributes like engine size, transmission type, and drivetrain configuration.

The dataset spans multiple vehicle classes, from compact cars to SUVs, and includes models from well-known manufacturers, with data ranging from older to newer models. By analyzing this data, we aim to uncover patterns and relationships that influence fuel efficiency, such as the impact of engine displacement, transmission type, etc on miles per gallon (mpg). Additionally, it can provide insights into trends over time, such as improvements in fuel efficiency or shifts in consumer preferences for vehicle types.

#### __*Problem Statement*__

In today’s world, fuel efficiency has become a top priority for customers who are looking to save on fuel costs and minimize their environmental footprint. The challenge lies in identifying the most fuel-efficient car models across various vehicle classes while uncovering the key factors that impact fuel economy. By tackling this issue, we can provide valuable insights to both manufacturers and consumers, helping them understand how different features, like engine size,car size and transmission affect fuel efficiency. This understanding can guide manufacturers in designing more efficient vehicles and empower consumers to make smarter, more eco-friendly purchasing decisions.

#### __*Objectives*__

- __Identify Fuel-Efficient Vehicles__

    - The goal is to identify the standout car models that excel in fuel efficiency. By ranking vehicles based on their combination MPG, the most eco-friendly options within each vehicle class can be uncovered, providing valuable insights for consumers looking to make smarter, greener choices..

- __Analyze Key Factors Influencing Fuel Efficiency__

    - This analysis will explore how key factors—such as engine displacement, the number of cylinders, drivetrain, transmission type, and fuel type—affect fuel efficiency. Understanding these elements will reveal the underlying factors that contribute to better fuel economy.

- __Class-Based Analysis__

    - Each vehicle class has its own efficiency leaders. By comparing the fuel efficiency of different types of vehicles, such as midsize, large cars, the top performers within each category will be highlighted, helping consumers choose the best option for their needs.

- __Explore Yearly Trends__

    - Examining data from 2014 to 2024, this analysis will track the progress of fuel efficiency, uncovering how technological advancements and design improvements have contributed to more fuel-efficient vehicles, and offering insights into the future of eco-friendly driving.   

---
## **Importing Packages**
<a id="one"></a>
<a href=#cont>Back to Table of Contents</a>



In [1]:
# importing packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

---

## **Data Collection and Description**
<a id="two"></a>
<a href=#cont>Back to Table of Contents</a>


#### __*Source*__

The dataset was collected from Kaggle, an online platform that hosts a variety of datasets shared by individuals, organizations, and data science communities. Kaggle datasets typically come from real-world scenarios and are often used for machine learning competitions, research, and educational purposes.

#### __*Description*__

Key Features:
- **city_mpg:** Fuel efficiency in miles per gallon (mpg) for city driving, crucial for urban driving analysis.
- **highway_mpg:** Fuel efficiency in mpg for highway driving, ideal for comparing long-distance performance.
- **combination_mpg:**** The combined fuel efficiency for both city and highway driving, a critical metric for overall vehicle performance.
- **cylinders & displacement:**** Engine specifications that help understand a car's power and fuel consumption.
- **drive:** The type of drivetrain (e.g., FWD, AWD), key for determining the car's handling and terrain suitability.
- **fuel_type:** The type of fuel (e.g., gasoline, electric), important for environmental impact studies.
- **make & model:** Car manufacturer and model, enabling brand-specific analysis and comparison.
- **year:** The production year of the vehicle, useful for tracking trends and technological improvements over time.

---
## **Data Loading and Inspection**
<a id="three"></a>
<a href=#cont>Back to Table of Contents</a>


The first step in the process involves loading the dataset into the analysis environment, ensuring that all necessary data is properly imported for further analysis

In [2]:
# Loading the data 
df = pd.read_csv('car_data.csv')

In [3]:
# Overview of the dataset
df.head()

,city_mpg,class,combination_mpg,cylinders,displacement,drive,fuel_type,highway_mpg,make,model,transmission,year
0,25,midsize car,29,4.0,2.5,fwd,gas,36,mazda,6,m,2014
1,26,midsize car,30,4.0,2.5,fwd,gas,37,mazda,6,a,2014
2,25,small sport utility vehicle,27,4.0,2.5,fwd,gas,31,mazda,cx-5 2wd,a,2014
3,26,small sport utility vehicle,29,4.0,2.0,fwd,gas,34,mazda,cx-5 2wd,m,2014
4,26,small sport utility vehicle,28,4.0,2.0,fwd,gas,32,mazda,cx-5 2wd,a,2014


In [4]:
# overview of structure and datatypes
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 550 entries, 0 to 549
Data columns (total 12 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   city_mpg         550 non-null    int64  
 1   class            550 non-null    object 
 2   combination_mpg  550 non-null    int64  
 3   cylinders        548 non-null    float64
 4   displacement     548 non-null    float64
 5   drive            550 non-null    object 
 6   fuel_type        550 non-null    object 
 7   highway_mpg      550 non-null    int64  
 8   make             550 non-null    object 
 9   model            550 non-null    object 
 10  transmission     550 non-null    object 
 11  year             550 non-null    int64  
dtypes: float64(2), int64(4), object(6)
memory usage: 51.7+ KB


The dataset consists of 550 entries and mix of numeric and object datatype, but some columns appear to contain missing values

In [5]:
# describing the dataset statistics
df.describe()

,city_mpg,combination_mpg,cylinders,displacement,highway_mpg,year
count,550.000000,550.000000,548.000000,548.000000,550.000000,550.000000
mean,21.460000,24.069091,5.315693,2.931752,28.609091,2019.000000
std,8.147392,7.478369,1.759999,1.248419,6.832228,3.165156
min,11.000000,14.000000,3.000000,1.200000,18.000000,2014.000000
25%,17.000000,20.000000,4.000000,2.000000,24.000000,2016.000000
50%,20.000000,23.000000,4.000000,2.500000,28.000000,2019.000000
75%,24.000000,27.000000,6.000000,3.500000,32.000000,2022.000000
max,126.000000,112.000000,12.000000,6.800000,102.000000,2024.000000


__Fuel Economy__

- The very high miles-per-gallon (MPG) values, such as 126 MPG for city driving and 112 MPG for combined driving might be outliers or it could indicate a possible mistake in data entry or the presence of vehicles designed for exceptional fuel efficiency, such as advanced hybrid cars.

__Engine Characteristics__

- Vehicles have between 3 to 12 cylinders, with a median of 4 cylinders, suggesting most cars are small to midsize.
- Engine displacement ranges from 1.2L to 6.8L, with a median of 2.5L, indicating a mix of economy and performance vehicles.

Aside from the extreme MPG values, most of the other values in the statistical summary table appear realistic

---
## **Data Cleaning and Filtering**
<a id="four"></a>
<a href=#cont>Back to Table of Contents</a>

* **Purpose:** Prepare the data for analysis by cleaning and filtering.
* **Details:** Include steps for handling missing values, removing outliers, correcting errors, and possibly reducing the data (filtering based on certain criteria or features).
---

In [ ]:
#Please use code cells to code in and do not forget to comment your code.

---

## **Exploratory Data Analysis (EDA)**
<a id="five"></a>
<a href=#cont>Back to Table of Contents</a>

* **Purpose:** Explore and visualize the data to uncover patterns, trends, and relationships.
* **Details:** Use statistics and visualizations to explore the data. This may include histograms, box plots, scatter plots, and correlation matrices. Discuss any significant findings.
---


In [ ]:
#Please use code cells to code in and do not forget to comment your code.

---

## **Modeling**
<a id="six"></a>
<a href=#cont>Back to Table of Contents</a>

* **Purpose:** Develop and train predictive or statistical models.
* **Details:** Describe the choice of models, feature selection and engineering processes, and show how the models are trained. Include code for setting up the models and explanations of the model parameters.
---


In [ ]:
#Please use code cells to code in and do not forget to comment your code.

---

## **Evaluation and Validation**
<a id="seven"></a>
<a href=#cont>Back to Table of Contents</a>

* **Purpose:** Evaluate and validate the effectiveness and accuracy of the models.
* **Details:** Present metrics used to evaluate the models, such as accuracy, precision, recall, F1-score, etc. Discuss validation techniques employed, such as cross-validation or train/test split.
---

In [ ]:
#Please use code cells to code in and do not forget to comment your code.

---

## **Final Model**
<a id="eight"></a>
<a href=#cont>Back to Table of Contents</a>

* **Purpose:** Present the final model and its performance.
* **Details:** Highlight the best-performing model and discuss its configuration, performance, and why it was chosen over others.
---


In [ ]:
#Please use code cells to code in and do not forget to comment your code.

---

## **Conclusion and Future Work**
<a id="nine"></a>
<a href=#cont>Back to Table of Contents</a>

* **Purpose:** Summarize the findings and discuss future directions.
* **Details:** Conclude with a summary of the results, insights gained, limitations of the study, and suggestions for future projects or improvements in methodology or data collection.
---


In [ ]:
#Please use code cells to code in and do not forget to comment your code.

---

## **References**
<a id="ten"></a>
<a href=#cont>Back to Table of Contents</a>

* **Purpose:** Provide citations and sources of external content.
* **Details:** List all the references and sources consulted during the project, including data sources, research papers, and documentation for tools and libraries used.
---

In [ ]:
#Please use code cells to code in and do not forget to comment your code.

## Additional Sections to Consider

* ### Appendix: 
For any additional code, detailed tables, or extended data visualizations that are supplementary to the main content.

* ### Contributors: 
If this is a group project, list the contributors and their roles or contributions to the project.
